BioC 2015: Where Software and Biology Connect
=============================================

Google has some pretty amazing big data computational "hammers" that they have been applying to search and video data for a long time.  In this workshop we take those same hammers and apply them to whole genome sequences.

We will work with both the :doc:`/use_cases/discover_public_data/1000_genomes` reads and variants and also the :doc:`/use_cases/discover_public_data/platinum_genomes` gVCF variants.

We do this all from the comfort of the R prompt using common packages including `VariantAnnotation`_, `ggbio`_, `ggplot2`_, `dplyr`_, `bigrquery`_, and the new Bioconductor package `GoogleGenomics`_ which provides an R interface to Google's implementation of the `Global Alliance for Genomics and Health API`_.

And we'll do this in a reproducible fashion running RMarkdown files via `Dockerized Bioconductor`_ running on `Google Compute Engine`_ VMs!

.. contents::

Get Started with Google Cloud Platform
--------------------------------------

Create a Google Cloud Platform project
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. include:: /includes/gcp_signup.rst

Enable APIs
^^^^^^^^^^^

Enable all the Google Cloud Platform APIs we will use in this workshop by clicking on this `link <https://console.developers.google.com/flows/enableapi?apiid=genomics,bigquery>`_.

Install gcloud
^^^^^^^^^^^^^^

.. include:: /includes/gcloud_setup.rst

Configure gcloud compute
^^^^^^^^^^^^^^^^^^^^^^^^

.. include:: /includes/gcloud_compute_configure.rst

Spin up a Google Compute Engine Instance running Bioconductor
-------------------------------------------------------------

This codelab `can be run from anywhere <https://github.com/googlegenomics/bioconductor-workshop-r>`_.  All the heavy lifting is happening in the cloud regardless of where R is running.  But if you like, you can run the whole darn thing in the cloud.

Bioconductor maintains Docker images with R, Bioconductor packages, and RStudio Server all ready to go!  Its a great way to set up your R environment quickly and start working.  The instructions are below but if you want to learn more, see http://www.bioconductor.org/help/docker/.

.. include:: /includes/r_workbench_setup.rst

You now have access to RStudio via your browser!

See https://github.com/googlegenomics/gce-images for the Docker file.  It depends upon http://www.bioconductor.org/help/docker/ which depends upon https://github.com/rocker-org/rocker/wiki.

.. container:: toggle

    .. container:: header

        If you prefer to setup R manually instead, click here to **Show/Hide Instructions**

    .. container:: content

      .. code::

        source("http://bioconductor.org/biocLite.R") # installs BiocInstaller
        useDevel() # specifies to use bioc-devel (3.2)
        biocLite("devtools") # needed for the special use of biocLite() below
        biocLite("googlegenomics/bioconductor-workshop-r", build_vignettes=TRUE, dependencies=TRUE)

Run the Codelabs
------------------------------------------------------

(1) View the workshop documentation.

.. code::

  help(package="Bioc2015Workshop")

(2) Click on *"User guides, package vignettes and other documentation."*

(3) Click on vignette *"Bioc2015Workshop"* and follow the instructions there to run the vignettes **line-by-line** or **chunk-by-chunk**!

  * To run **line-by-line**, put your cursor on the desired line  and click the "Run" button or use keyboard shortcuts for Windows/Linux: ``Ctrl+Enter`` and Mac: ``Command+Enter``.

  * To run **chunk-by-chunk**, put your cursor in the desired chunk and click the "Chunks -> Run Current Chuck" button. or use keyboard shortcuts for Windows/Linux: ``Ctrl+Alt+C`` and Mac: ``Command+Option+C``.

  .. image:: /_static/run_rmarkdown.png
    :alt: Run Rmarkdown

Looking for more to try?
-----------------
* Try these samples on different datasets :doc:`/use_cases/discover_public_data/index`.
* Find more example BigQuery queries in:

  * `getting-started-bigquery`_
  * `bigquery-examples`_
  * `Quality Control using Google Genomics`_

* Run a `Google Cloud Dataflow`_ pipeline:

  * :doc:`/use_cases/compute_identity_by_state/index`
  * :doc:`/use_cases/compute_principal_coordinate_analysis/1-way-pca`

Wrap up
-------------

"Stop" or "Delete" your VM
^^^^^^^^^^^^^^^^^^^^^^^^^^

(1) Go to the Google Developers Console: https://console.developers.google.com/project/_/compute/instances

(2) Click on the checkbox next to your VM.

(3) Click on "Stop" to pause your VM or "Delete" to delete it.

For more detail, see: https://cloud.google.com/compute/docs/instances/#stop_job

Stay Involved
^^^^^^^^^^^^^

* Collaborate on the `GA4GH`_ specification.
* :doc:`/mailinglist`.
* Collaborate on https://github.com/Bioconductor/GoogleGenomics.
